### ISSUE TITLE
Dokku do not kill the old resque container

### ISSUE THREAD

**Issue by phuongnd08** [dokku/dokku#2979]
Description of problem:
I have a Procfile that define web, resque and a scheduler process.
When a new deploy happen, dokku doesn't kill the old containers, resulting in multiple container of same process being run, all old and new.

My `dokku report`
```
<!-- If you're running 0.4.10+ just use dokku report-->
-----> uname: Linux ip-10-0-0-215 4.4.0-1038-aws #47-Ubuntu SMP Thu Sep 28 20:05:35 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
-----> memory:
              total        used        free      shared  buff/cache   available
Mem:           7982         928        3689          17        3365        6660
Swap:             0           0           0
-----> docker version:
Client:
 Version:      17.10.0-ce
 API version:  1.33
 Go version:   go1.8.3
 Git commit:   f4ffd25
 Built:        Tue Oct 17 19:04:16 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.10.0-ce
 API version:  1.33 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   f4ffd25
 Built:        Tue Oct 17 19:02:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
-----> docker daemon info:
Containers: 20
 Running: 6
 Paused: 0
 Stopped: 14
Images: 43
Server Version: 17.10.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 0351df1c5a66838d0c392b4ac4cf9450de844e2d
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-1038-aws
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.795GiB
Name: ip-10-0-0-215
ID: 6VTY:PY3Z:U6IB:XJ3S:LZAF:SSXG:7NC7:DUUI:VVT3:7O7X:SAPU:SIZH
Docker Root Dir: /var/lib/docker
Debug Mode (client): true
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
-----> sigil version: 0.4.0
-----> herokuish version:
herokuish: 0.3.32
buildpacks:
  heroku-buildpack-multi     v1.0.0
  heroku-buildpack-ruby      v168
  heroku-buildpack-nodejs    v99
  heroku-buildpack-clojure   v77
  heroku-buildpack-python    v99
  heroku-buildpack-java      v53
  heroku-buildpack-gradle    v23
  heroku-buildpack-scala     v78
  heroku-buildpack-play      v26
  heroku-buildpack-php       v122
  heroku-buildpack-go        v72
  heroku-buildpack-erlang    fa17af9
  buildpack-nginx            v8
-----> dokku version: 0.10.5
-----> dokku plugins:
plugn: 0.3.0
  00_dokku-standard    0.10.5 enabled    dokku core standard plugin
  20_events            0.10.5 enabled    dokku core events logging plugin
  apps                 0.10.5 enabled    dokku core apps plugin
  build-env            0.10.5 enabled    dokku core build-env plugin
  certs                0.10.5 enabled    dokku core certificate management plugin
  checks               0.10.5 enabled    dokku core checks plugin
  common               0.10.5 enabled    dokku core common plugin
  config               0.10.5 enabled    dokku core config plugin
  deployment-keys      0.2.0 enabled    Manage SSH deployment keys that should get injected into your containers on-build
  docker-options       0.10.5 enabled    dokku core docker-options plugin
  domains              0.10.5 enabled    dokku core domains plugin
  enter                0.10.5 enabled    dokku core enter plugin
  git                  0.10.5 enabled    dokku core git plugin
  hostkeys-plugin      0.2.1 enabled    Manage hostkeys (.ssh/known_hosts) in your container environment
  logs                 0.10.5 enabled    dokku core logs plugin
  named-containers     0.10.5 enabled    dokku core named containers plugin
  nginx-vhosts         0.10.5 enabled    dokku core nginx-vhosts plugin
  plugin               0.10.5 enabled    dokku core plugin plugin
  proxy                0.10.5 enabled    dokku core proxy plugin
  ps                   0.10.5 enabled    dokku core ps plugin
  repo                 0.10.5 enabled    dokku core repo plugin
  shell                0.10.5 enabled    dokku core shell plugin
  ssh-keys             0.10.5 enabled    dokku core ssh-keys plugin
  storage              0.10.5 enabled    dokku core storage plugin
  tags                 0.10.5 enabled    dokku core tags plugin
  tar                  0.10.5 enabled    dokku core tar plugin
```

Procfile:
```
web: bundle exec thin start -p $PORT
resque: env TERM_CHILD=1 RESQUE_TERM_TIMEOUT=7 bundle exec resque-pool
clockwork: bundle exec clockwork clockwork.rb
```

Environment details (AWS, VirtualBox, physical, etc.):
AWS Ubuntu 16.04

How was Dokku installed?:
```
wget https://raw.githubusercontent.com/dokku/dokku/v0.10.5/bootstrap.sh
sudo DOKKU_TAG=v0.10.5 bash bootstrap.sh
```
How reproducible:

Steps to Reproduce:
1. Create a dokku app
2. Create a rails app with resque
3. Push the rails app to dokku, scale resque=1
4. Make some changes to rails app, push again within 20 seconds

Actual Results:
You have 2 running resque containers

Expected Results:
You have only 1 running resque containers

**Comment by phuongnd08** [comment]
Look like it was an issue with restarting policies. After changing restarting policies to no, I can see the old resque being killed.

**Comment by phuongnd08** [comment]
Okay, now it's clear that if I make successive deploys without waiting for dokku to finish cleanup, there will be some obsolete containers left running. Probably because dokku clean up acts on a shared list of containers which can be overridden by new deploy?

**Comment by josegonzalez** [comment]
When we deploy, we leave old containers running for one minute. This is to let old processes drain pending requests.

You can manage this via the checks plugin and some environment variables as [documented here](http://dokku.viewdocs.io/dokku/deployment/zero-downtime-deploys/).

Closing as this is working as designed.

**Comment by phuongnd08** [comment]
Running for a minute is fine. What I'm seeing here is that container left running for many hours until I kill them with docker.

**Comment by josegonzalez** [comment]
We actually don't keep "track" of old containers in an index or something. Instead, we fork something in the background to do this, as shown [here](https://github.com/dokku/dokku/blob/master/plugins/common/functions#L672).

Can you verify that the forked process is still there, or that we're not triggering the `retire-container-failed` plugin hook?

**Comment by phuongnd08** [comment]
I'm currently avoiding that by setting the retire time to zero seconds and avoiding triggering 2 deploy/scale command in parallel and so far it has been reliable. I'll extract these information next time. Btw how do I know if it's triggering retire-container-failed plugin?

Also keeping background worker running for a minute may be risky. Certain process may be designed to not running in parallel (like worker that process transaction). So I'd rather retire these containers immediately.

**Comment by josegonzalez** [comment]
You can turn on the [event log](http://dokku.viewdocs.io/dokku/advanced-usage/event-logs/), which will log every plugin trigger to a file.

Regarding background workers being retired at a different length of time, sure. We don't have that option, mostly because no one has asked/worked on it. I think thats a separate conversation we should have, and figure out what the correct interface to surface such configuration. At the moment, you can set the time globally and then per app, but not per app/worker.

Lets first figure out why this isn't working as expected for you before getting to any feature enhancements (which should be in a separate issue).

**Comment by josegonzalez** [comment]
Closing as there hasn't been any update in three weeks. Feel free to comment with more info and we'll be glad to take a look.

**Comment by Loschcode** [comment]
This issue still happens. I deployed 2 apps at the same time (api / app from the same project) and my Sidekiq / Redis wouldn't take the new codebase into consideration. I figured that might be a problem of residual container or something like that.

I had to remove my containers manually and rebuild all the apps of my droplet to fix the problem; no server restart or stop from Dokku would fix it. That sounds pretty critical to me. I'm on Dokku 0.17.5.
